File: dashboard/qa-evidence/run-l2-qa.py
import subprocess
import json
import time
import os
import sys

EVIDENCE_DIR = '/Users/wubaizong/接案/ZenOS/dashboard/qa-evidence'
os.makedirs(EVIDENCE_DIR, exist_ok=True)

# Generate token
token_result = subprocess.run(
    ['node', 'scripts/gen-test-token.js'],
    capture_output=True, text=True,
    cwd='/Users/wubaizong/接案/ZenOS/dashboard'
)
TEST_TOKEN = token_result.stdout.strip()
print(f'Token (first 50): {TEST_TOKEN[:50]}...')

from playwright.sync_api import sync_playwright

results = {}

def screenshot(page, name, note=''):
    fp = os.path.join(EVIDENCE_DIR, f'{name}.png')
    page.screenshot(path=fp, full_page=False)
    print(f'  Screenshot: {fp} {note}')
    return fp

with sync_playwright() as p:
    browser = p.chromium.launch(headless=True, slow_mo=200)
    context = browser.new_context(viewport={'width': 1440, 'height': 900})
    page = context.new_page()
    
    try:
        # ═══════════════════════════════════════════════
        # T1: Initial State
        # ═══════════════════════════════════════════════
        print('\n=== T1: Initial State ===')
        page.goto('http://localhost:3000/knowledge-map')
        page.wait_for_timeout(3000)  # Wait for Firebase SDK
        
        # Inject custom token
        sign_result = page.evaluate("""async (token) => {
            const signIn = window.__signInWithCustomToken;
            if (!signIn) return { error: '__signInWithCustomToken not found' };
            try {
                await signIn(token);
                return { ok: true };
            } catch(e) {
                return { error: e.message };
            }
        }""", TEST_TOKEN)
        print(f'  Sign in result: {sign_result}')
        
        # Wait for page to load after auth
        try:
            page.wait_for_load_state('networkidle', timeout=15000)
        except:
            pass
        page.wait_for_timeout(3000)
        
        screenshot(page, 'T1-initial-state')
        
        # Verify: canvas is visible, no error
        canvas = page.locator('canvas').first()
        canvas_visible = canvas.is_visible()
        print(f'  Canvas visible: {canvas_visible}')
        
        # Check for "Loading" text gone
        body_text = page.inner_text('body')
        loading_gone = 'Loading knowledge map' not in body_text
        print(f'  Loading gone: {loading_gone}')
        print(f'  Body preview: {body_text[:200]}')
        
        has_login_error = sign_result.get('error') is not None
        t1_pass = canvas_visible and not has_login_error
        results['T1'] = {'pass': t1_pass, 'note': f"Canvas:{canvas_visible}, Login:{sign_result}"}
        print(f'  T1: {"PASS" if t1_pass else "FAIL"}')
        
        # ═══════════════════════════════════════════════
        # T2: Click L2 module node → expand + popover
        # ═══════════════════════════════════════════════
        print('\n=== T2: Click L2 module node ===')
        
        # Wait for graph to settle
        page.wait_for_timeout(3000)
        screenshot(page, 'T2a-before-click')
        
        canvas_bb = canvas.bounding_box()
        print(f'  Canvas BB: {canvas_bb}')
        
        if canvas_bb:
            cx = canvas_bb['x'] + canvas_bb['width'] / 2
            cy = canvas_bb['y'] + canvas_bb['height'] / 2
        else:
            cx, cy = 720, 450
        
        print(f'  Canvas center: ({cx}, {cy})')
        
        # Try multiple positions to find module node (purple L2)
        # Graph typically has L1 product in center, L2 modules spread around
        positions = [
            (cx, cy),
            (cx - 120, cy - 90),
            (cx + 120, cy - 90),
            (cx - 120, cy + 90),
            (cx + 120, cy + 90),
            (cx, cy - 150),
            (cx, cy + 150),
            (cx - 180, cy),
            (cx + 180, cy),
            (cx - 200, cy - 130),
            (cx + 200, cy - 130),
            (cx - 200, cy + 130),
            (cx + 200, cy + 130),
        ]
        
        module_clicked = False
        popover_visible = False
        clicked_pos = None
        
        for x, y in positions:
            print(f'  Trying click at ({x:.0f}, {y:.0f})...')
            page.mouse.click(x, y)
            page.wait_for_timeout(1200)
            
            # Check for checkboxes (popover appeared)
            checkbox_count = page.locator('input[type="checkbox"]').count()
            if checkbox_count > 0:
                print(f'  Checkboxes found: {checkbox_count} — popover opened!')
                module_clicked = True
                popover_visible = True
                clicked_pos = (x, y)
                break
            
            # Check for z-50 class (popover div)
            z50_count = page.locator('.z-50').count()
            if z50_count > 0:
                z50_visible = page.locator('.z-50').first().is_visible()
                if z50_visible:
                    print(f'  z-50 element visible — popover!')
                    module_clicked = True
                    popover_visible = True
                    clicked_pos = (x, y)
                    break
        
        screenshot(page, 'T2-after-click')
        print(f'  Module clicked: {module_clicked}, Popover: {popover_visible}, pos: {clicked_pos}')
        
        results['T2'] = {
            'pass': module_clicked and popover_visible,
            'note': f"Module:{module_clicked}, Popover:{popover_visible}, pos:{clicked_pos}"
        }
        print(f'  T2: {"PASS" if results["T2"]["pass"] else "FAIL"}')
        
        # ═══════════════════════════════════════════════
        # T3: Checklist popover content
        # ═══════════════════════════════════════════════
        print('\n=== T3: Checklist popover content ===')
        has_document = False
        has_task = False
        checkbox_info = []
        
        if popover_visible:
            # Get all visible z-50 content
            z50_text = ''
            try:
                z50_text = page.locator('.z-50').first().inner_text()
                print(f'  Popover text: {z50_text}')
                has_document = 'document' in z50_text.lower() or 'Document' in z50_text
                has_task = 'task' in z50_text.lower() or 'Task' in z50_text
            except Exception as e:
                print(f'  z-50 text error: {e}')
            
            # Check individual checkboxes
            checkboxes = page.locator('input[type="checkbox"]')
            cb_count = checkboxes.count()
            for i in range(cb_count):
                cb = checkboxes.nth(i)
                label = cb.evaluate("""el => {
                    const parent = el.closest('label') || el.parentElement;
                    return parent ? parent.innerText.trim() : '';
                }""")
                checked = cb.is_checked()
                print(f'  Checkbox {i}: "{label}" checked={checked}')
                checkbox_info.append({'label': label, 'checked': checked})
                if 'document' in label.lower():
                    has_document = True
                if 'task' in label.lower():
                    has_task = True
        
        screenshot(page, 'T3-checklist-popover')
        t3_pass = popover_visible and has_document and has_task
        results['T3'] = {
            'pass': t3_pass,
            'note': f"Document:{has_document}, Task:{has_task}, Checkboxes:{checkbox_info}"
        }
        print(f'  T3: {"PASS" if t3_pass else "FAIL"}')
        
        # ═══════════════════════════════════════════════
        # T4: Uncheck Task → task nodes disappear
        # ═══════════════════════════════════════════════
        print('\n=== T4: Uncheck Task ===')
        task_unchecked = False
        
        if popover_visible:
            checkboxes = page.locator('input[type="checkbox"]')
            cb_count = checkboxes.count()
            print(f'  Total checkboxes: {cb_count}')
            
            for i in range(cb_count):
                cb = checkboxes.nth(i)
                label = cb.evaluate("""el => {
                    const parent = el.closest('label') || el.parentElement;
                    return parent ? parent.innerText.trim() : '';
                }""")
                is_checked = cb.is_checked()
                print(f'  Checkbox {i}: "{label}" checked={is_checked}')
                
                if 'task' in label.lower() and is_checked:
                    print(f'  Unchecking Task checkbox...')
                    cb.click()
                    page.wait_for_timeout(800)
                    still_checked = cb.is_checked()
                    print(f'  After click, checked: {still_checked}')
                    task_unchecked = not still_checked
                    break
            
            # Fallback: try clicking on "Task" text
            if not task_unchecked:
                try:
                    task_labels = page.get_by_text('Task', exact=True)
                    tc = task_labels.count()
                    print(f'  "Task" text elements: {tc}')
                    if tc > 0:
                        task_labels.first().click()
                        page.wait_for_timeout(800)
                        task_unchecked = True
                        print('  Clicked Task text')
                except Exception as e:
                    print(f'  Task text click error: {e}')
        
        screenshot(page, 'T4-task-unchecked')
        results['T4'] = {
            'pass': task_unchecked,
            'note': f"Task unchecked: {task_unchecked}"
        }
        print(f'  T4: {"PASS" if task_unchecked else "FAIL"}')
        
        # ═══════════════════════════════════════════════
        # T5: Click same L2 again → collapse
        # ═══════════════════════════════════════════════
        print('\n=== T5: Click same L2 again → collapse ===')
        
        # Close popover by pressing Escape first
        page.keyboard.press('Escape')
        page.wait_for_timeout(500)
        
        # Check initial checkbox count
        cb_before = page.locator('input[type="checkbox"]').count()
        print(f'  Checkboxes before re-click: {cb_before}')
        
        # Re-click the position where module was found
        if clicked_pos:
            x, y = clicked_pos
            print(f'  Re-clicking at ({x:.0f}, {y:.0f})...')
            page.mouse.click(x, y)
            page.wait_for_timeout(1500)
            
            cb_after = page.locator('input[type="checkbox"]').count()
            print(f'  Checkboxes after re-click: {cb_after}')
            
            # If re-click opened popover again, need to click once more to collapse
            if cb_after > 0:
                print('  Popover reopened — need another click to collapse')
                page.mouse.click(x, y)
                page.wait_for_timeout(1500)
                cb_final = page.locator('input[type="checkbox"]').count()
                print(f'  Checkboxes after 2nd click: {cb_final}')
                collapse_happened = cb_final == 0
            else:
                # Module was collapsed
                collapse_happened = cb_after == 0
        else:
            collapse_happened = False
        
        screenshot(page, 'T5-after-collapse')
        results['T5'] = {
            'pass': collapse_happened,
            'note': f"Collapsed: {collapse_happened}"
        }
        print(f'  T5: {"PASS" if collapse_happened else "FAIL"}')
        
    except Exception as e:
        print(f'ERROR: {e}')
        import traceback
        traceback.print_exc()
        screenshot(page, 'error-state')
    finally:
        screenshot(page, 'final-state')
        browser.close()

print('\n' + '='*50)
print('RESULTS SUMMARY')
print('='*50)
for test, result in results.items():
    status = 'PASS' if result['pass'] else 'FAIL'
    print(f'{test}: {status} — {result["note"]}')

all_p0_pass = all(results.get(t, {}).get('pass', False) for t in ['T1','T2','T3','T4','T5'])
print(f'\nOverall P0: {"ALL PASS" if all_p0_pass else "SOME FAILED"}')

File: dashboard/qa-evidence/run-l2-qa.js
const { chromium } = require('playwright');
const path = require('path');
const { execSync } = require('child_process');
const fs = require('fs');

// Generate token
const token = execSync('node /Users/wubaizong/接案/ZenOS/dashboard/scripts/gen-test-token.js').toString().trim();
console.log('Token generated (first 50 chars):', token.substring(0, 50) + '...');

const EVIDENCE_DIR = '/Users/wubaizong/接案/ZenOS/dashboard/qa-evidence';

async function screenshot(page, name) {
  const fp = path.join(EVIDENCE_DIR, `${name}.png`);
  await page.screenshot({ path: fp, fullPage: false });
  console.log(`Screenshot: ${fp}`);
  return fp;
}

async function main() {
  const browser = await chromium.launch({ headless: false, slowMo: 300 });
  const context = await browser.newContext({ viewport: { width: 1440, height: 900 } });
  const page = await context.newPage();

  const results = {};

  try {
    // ── Step 1: Navigate to knowledge-map
    console.log('\n=== T1: Initial State ===');
    await page.goto('http://localhost:3000/knowledge-map');
    await page.waitForTimeout(3000); // Wait for Firebase SDK

    // Inject custom token
    console.log('Injecting custom token...');
    const signInResult = await page.evaluate(async (token) => {
      const signIn = window.__signInWithCustomToken;
      if (!signIn) return { error: '__signInWithCustomToken not found' };
      try {
        await signIn(token);
        return { ok: true };
      } catch(e) {
        return { error: e.message };
      }
    }, token);
    console.log('Sign in result:', signInResult);

    if (signInResult.error) {
      results.T1 = { pass: false, note: 'Login hook not found: ' + signInResult.error };
    }

    // Wait for page to update after auth
    await page.waitForLoadState('networkidle', { timeout: 15000 });
    await page.waitForTimeout(2000);

    // T1 screenshot
    await screenshot(page, 'T1-initial-state');

    // Check for graph canvas
    const canvas = page.locator('canvas').first();
    const canvasVisible = await canvas.isVisible().catch(() => false);
    console.log('Canvas visible:', canvasVisible);

    // Check for error indicators
    const bodyText = await page.locator('body').innerText().catch(() => '');
    const hasError = bodyText.includes('Error') || bodyText.includes('error');
    console.log('Body has error text:', hasError);
    console.log('Body preview:', bodyText.substring(0, 200));

    results.T1 = {
      pass: canvasVisible && !signInResult.error,
      note: `Canvas visible: ${canvasVisible}, Login: ${signInResult.error ?? 'ok'}`
    };

    // ── T2: Click L2 module node
    console.log('\n=== T2: Click L2 module node ===');
    // Get canvas bounding box
    const canvasBB = await canvas.boundingBox();
    console.log('Canvas bounding box:', canvasBB);

    // Wait for graph simulation to settle
    await page.waitForTimeout(3000);
    await screenshot(page, 'T2a-before-click');

    // Try to find module nodes by examining the page state
    // We'll query exposed node positions via JS if available
    const nodePositions = await page.evaluate(() => {
      // Try to get from window debug state
      if (window.__graphNodes) return window.__graphNodes;
      return null;
    });
    console.log('Graph nodes from window:', nodePositions);

    // Strategy: click near center of canvas (where graph usually is)
    // Graph typically renders nodes spread around center
    const cx = canvasBB.x + canvasBB.width / 2;
    const cy = canvasBB.y + canvasBB.height / 2;
    console.log(`Canvas center: (${cx}, ${cy})`);

    // Try multiple positions to find a module node
    // Modules (L2, purple) are children of product, spread around center
    const tryPositions = [
      [cx, cy],                          // dead center
      [cx - 100, cy - 80],              // top-left area
      [cx + 100, cy - 80],              // top-right area
      [cx - 100, cy + 80],              // bottom-left area  
      [cx + 100, cy + 80],              // bottom-right area
      [cx, cy - 120],                   // top
      [cx, cy + 120],                   // bottom
      [cx - 150, cy],                   // left
      [cx + 150, cy],                   // right
    ];

    let moduleClicked = false;
    let popoverVisible = false;
    let newNodesAppeared = false;

    for (const [x, y] of tryPositions) {
      console.log(`Trying click at (${x}, ${y})`);
      await page.mouse.click(x, y);
      await page.waitForTimeout(1500);

      // Check if popover appeared
      const popover = page.locator('[class*="popover"], [class*="TypeFilter"], .z-50').first();
      popoverVisible = await popover.isVisible().catch(() => false);
      
      if (popoverVisible) {
        console.log('Popover appeared at position:', x, y);
        moduleClicked = true;
        break;
      }
      
      // Also check if any checklist-like content appeared
      const checkboxes = page.locator('input[type="checkbox"]');
      const checkboxCount = await checkboxes.count().catch(() => 0);
      if (checkboxCount > 0) {
        console.log('Checkboxes appeared, count:', checkboxCount);
        moduleClicked = true;
        popoverVisible = true;
        break;
      }
    }

    await screenshot(page, 'T2-after-click');
    console.log('Module clicked (popover appeared):', moduleClicked);

    results.T2 = {
      pass: moduleClicked && popoverVisible,
      note: `Module click: ${moduleClicked}, Popover: ${popoverVisible}`
    };

    // ── T3: Checklist popover content
    console.log('\n=== T3: Checklist popover content ===');
    let checkboxes = [];
    let hasDocument = false;
    let hasTask = false;

    if (popoverVisible) {
      const checkboxEls = page.locator('input[type="checkbox"]');
      const count = await checkboxEls.count().catch(() => 0);
      console.log('Checkbox count:', count);

      for (let i = 0; i < count; i++) {
        const cb = checkboxEls.nth(i);
        const label = await cb.evaluate(el => {
          const parent = el.closest('label') || el.parentElement;
          return parent ? parent.innerText : '';
        }).catch(() => '');
        const checked = await cb.isChecked().catch(() => false);
        console.log(`Checkbox ${i}: "${label}" checked=${checked}`);
        checkboxes.push({ label, checked });
        if (label.toLowerCase().includes('document')) hasDocument = true;
        if (label.toLowerCase().includes('task')) hasTask = true;
      }

      // Also check for text labels near checkboxes
      const popoverText = await page.locator('.z-50').innerText().catch(() => '');
      console.log('Popover text:', popoverText);
      if (popoverText.toLowerCase().includes('document')) hasDocument = true;
      if (popoverText.toLowerCase().includes('task')) hasTask = true;
    }

    await screenshot(page, 'T3-checklist-popover');

    results.T3 = {
      pass: popoverVisible && hasDocument && hasTask,
      note: `Has Document: ${hasDocument}, Has Task: ${hasTask}, Checkboxes: ${checkboxes.length}`
    };

    // ── T4: Uncheck Task → task nodes disappear
    console.log('\n=== T4: Uncheck Task ===');
    let taskUnchecked = false;

    if (popoverVisible) {
      // Find Task checkbox
      const allCheckboxes = page.locator('input[type="checkbox"]');
      const count = await allCheckboxes.count().catch(() => 0);
      
      for (let i = 0; i < count; i++) {
        const cb = allCheckboxes.nth(i);
        const label = await cb.evaluate(el => {
          const parent = el.closest('label') || el.parentElement;
          return parent ? parent.innerText : '';
        }).catch(() => '');
        const isChecked = await cb.isChecked().catch(() => false);
        
        if (label.toLowerCase().includes('task') && isChecked) {
          console.log('Unchecking Task checkbox...');
          await cb.click();
          await page.waitForTimeout(1000);
          taskUnchecked = true;
          break;
        }
      }

      // If direct checkbox didn't work, try clicking on label text
      if (!taskUnchecked) {
        const taskLabel = page.getByText('Task').first();
        const visible = await taskLabel.isVisible().catch(() => false);
        if (visible) {
          await taskLabel.click();
          await page.waitForTimeout(1000);
          taskUnchecked = true;
        }
      }
    }

    await screenshot(page, 'T4-task-unchecked');
    console.log('Task unchecked:', taskUnchecked);

    results.T4 = {
      pass: taskUnchecked,
      note: `Task checkbox unchecked: ${taskUnchecked}`
    };

    // ── T5: Click same L2 again → collapse
    console.log('\n=== T5: Click same L2 again → collapse ===');
    // Close popover first by pressing Escape or clicking elsewhere
    await page.keyboard.press('Escape');
    await page.waitForTimeout(500);

    // Re-click the same position where we found the module
    let collapseClicked = false;
    for (const [x, y] of tryPositions) {
      await page.mouse.click(x, y);
      await page.waitForTimeout(1500);
      
      // Check if popover disappeared (collapsed)
      const checkboxCount = await page.locator('input[type="checkbox"]').count().catch(() => 0);
      if (checkboxCount === 0) {
        console.log('No checkboxes visible — collapsed!');
        collapseClicked = true;
        break;
      }
    }

    await screenshot(page, 'T5-after-collapse');
    console.log('Collapse happened:', collapseClicked);

    results.T5 = {
      pass: collapseClicked,
      note: `Collapsed: ${collapseClicked}`
    };

  } catch (err) {
    console.error('Test error:', err);
  } finally {
    await screenshot(page, 'final-state');
    await browser.close();

    console.log('\n=== RESULTS SUMMARY ===');
    for (const [test, result] of Object.entries(results)) {
      const status = result.pass ? 'PASS' : 'FAIL';
      console.log(`${test}: ${status} — ${result.note}`);
    }

    const allP0Pass = ['T1','T2','T3','T4','T5'].every(t => results[t]?.pass);
    console.log('\nOverall P0:', allP0Pass ? 'ALL PASS' : 'SOME FAILED');
  }
}

main().catch(console.error);
